x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/NativeMemoryCalculator.java

```diff
@@ -199,22 +199,25 @@ public static long dynamicallyCalculateJvmSizeFromNativeMemorySize(long nativeMa
     public static ByteSizeValue calculateMaxModelMemoryLimitToFit(ClusterSettings clusterSettings, DiscoveryNodes nodes) {
 
         long maxMlMemory = 0;
-        int numMlNodes = 0;
 
         for (DiscoveryNode node : nodes) {
             OptionalLong limit = allowedBytesForMl(node, clusterSettings);
             if (limit.isEmpty()) {
                 continue;
             }
             maxMlMemory = Math.max(maxMlMemory, limit.getAsLong());
-            ++numMlNodes;
         }
 
         // It is possible that there is scope for more ML nodes to be added
         // to the cluster, in which case take those into account too
         long maxMlNodeSize = clusterSettings.get(MAX_ML_NODE_SIZE).getBytes();
         int maxLazyNodes = clusterSettings.get(MAX_LAZY_ML_NODES);
-        if (maxMlNodeSize > 0 && numMlNodes < maxLazyNodes) {
+        // Even if all the lazy nodes have been added to the cluster, we make
+        // the assumption that if any were configured they'll be able to grow
+        // to the maximum ML node size. (We assuming that lazy nodes always
+        // behave like they do with Elastic Cloud autoscaling, where vertical
+        // scaling is possible.)
+        if (maxMlNodeSize > 0 && maxLazyNodes > 0) {
```

Comment:
The fix here is consistent with the way #84242 was fixed.  That was a similar problem.